github.go

```diff
@@ -13,9 +13,17 @@ type GitHubRepo struct {
 	Name  string // The GitHub repo name
 }
 
+// Represents a specific git commit.
+// Note that code using GitHub Commit should respect the following hierarchy:
+// - CommitSha > BranchName > GitTag
+// - Example: GitTag and BranchName are both specified; use the GitTag
+// - Example: GitTag and CommitSha are both specified; use the CommitSha
+// - Example: BranchName alone is specified; use BranchName
 type GitHubCommit struct {
-	repo   GitHubRepo // The GitHub repo where this release lives
-	gitTag string     // The specific git tag for this release
+	Repo       GitHubRepo // The GitHub repo where this release lives
+	GitTag     string     // The specific git tag for this release
+	BranchName string     // If specified, indicates that this commit should be the latest commit on the given branch
+	CommitSha  string     // If specified, indicates that this commit should be exactly this Git Commit SHA.
 }
```

Comment:
Whenever I see a pattern where you have a type where a number of the fields may be empty, and you have if-statements all over the code to check for those empty values, it usually means that you can model the data a bit cleaner to avoid that. 

For example, you could make GitHubCommit an `interface`:

```go
type GitHubCommit interface {
    Repo() GitHubRepo
    Url() string
    Description() string
}
```

You can then define different structs that only define the fields that should actually be set, plus the corresponding methods to implement the `GitHubCommit` interface. For example, here is the one for when a user specifies a Tag:

```go
type GitHubTag struct {
    Repo GitHubRepo
    GitTag string
}

func (tag *GitHubTag) GitHubRepo() GitHubRepo {
    return tag.Repo
}

func (tag *GitHubTag) Url() string {
    return fmt.Sprintf("https://api.github.com/repos/%s/%s/zipball/%s", tag.Repo.Owner, tag.Repo.Name, tag.GitTag)
} 

func (tag *GitHubTag) Description() string {
    return fmt.Sprintf("Tag \"%s\" of %s ...\n", tag.GitTag, tag.Repo.Url)
}
```

Now, whenever you create a `GitHubTag` object, you always expect the `Tag` field to be set. Moreover, you use known methods to get its URL, description, etc. instead of if-statements all over the place.